src/ShellClient.ts

```diff
@@ -1,14 +1,91 @@
 
-export class ShellCleint {
+export class ShellClient {
 
-  constructor() {
-    // dummy constructor implementation for build testing
-    console.log('constructing new ShellClient ');
+  private static _instance: ShellClient;
+
+  private postMessageHandler: ((type: string, value: any) => void) | undefined;
+
+  private subscribersMap: Map<string, Function[]>
+
+  private constructor(
+    private target: Window,
+    private origin: string
+  ) {
+    this.subscribersMap = new Map();
+    this.initMessageApi();
+  }
+
+  public static init(target: Window, origin: string): ShellClient {
+    if (!ShellClient._instance) {
+      ShellClient._instance = new ShellClient(target, origin);
+    }
+    return ShellClient.instance;
+  }
+
+  public static get instance(): ShellClient {
+    if (!ShellClient._instance) {
+      throw new Error('ShellClient wasn\'t initialized.');
+    }
+    return ShellClient._instance;
+  }
+
+  public setTarget(target: Window, origin: string) {
+    const targetChanged = this.target !== target || this.origin !== origin;
+    if (targetChanged) {
+      this.target = target;
+      this.origin = origin;
+    }
+  }
+
+
+  public on = (type: string, subscriber: Function): Function => {
```

Comment:
can we have a `eventType`-type for the `type` here so we get autocomplete by using it by default.
so 
```
type eventType = 'V1.REQUIRE_CONTEXT' | string
```  
for example.